### ISSUE TITLE
Make `pub` in a binary be equivalent to `pub(crate)`

### ISSUE THREAD

**Issue by jhpratt** [rust-lang/rust#74970]
<sup>(copy-pasting from [IRLO](https://internals.rust-lang.org/t/should-pub-in-a-binary-be-equivalent-to-pub-crate/12821))</sup>

Consider the code snippet

```rust
pub fn foo() {
    // unused function!
}

fn main() {}
```

`cargo check` reports no errors or warnings. However, the same snippet using `pub(crate) fn foo()` gives a warning:

```text
warning: function is never used: `foo`
 --> src/main.rs:1:15
  |
1 | pub(crate) fn foo() {
  |               ^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: 1 warning emitted
```

Given that binaries can't be used in anything else (as far as I'm aware?), would it make sense to treat `pub` as equivalent to `pub(crate)` in binaries? I don't believe there is any difference in semantics, but it certainly makes a difference regarding warnings.

---

CC @estebank, who suggested I file an issue.

**Comment by bjorn3** [comment]
> Given that binaries can't be used in anything else

A dylib loaded by the binary could lookup the symbol using `dlsym` if it was `#[no_mangle]` I think.

**Comment by jonas-schievink** [comment]
You can build multiple crate types at once. Then this would break any non-binary crate types that are built at the same time.

What about just changing the lint instead? It could check if the only crate type being built is a binary, and then treat only `fn main` as the initially used function.

**Comment by Nemo157** [comment]
```
the target `foo` is a binary and can't have any crate-types set (currently "lib")
```

So this could only apply to _binary targets_, not the crate-type `bin` (if you have a library target that is set to be something like `["bin", "staticlib"]`).

**Comment by jhpratt** [comment]
Changing the lint would certainly be sensible if totally merging the behavior causes issues. I believe this affects any item, not just functions. It would probably be worth checking whether any other lints have the same type of behavior.

**Comment by daboross** [comment]
> A dylib loaded by the binary could lookup the symbol using `dlsym` if it was `#[no_mangle]` I think.

In this case, `#[no_mangle]` itself should mark the function as used, so I don't think this would present any problems. The following library compiles without warnings on stable rustc:

```rust
#[no_mangle]
fn foo() {
}
```

([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dd1001dc870f59c2274d0d80636ccb7b))

**Comment by ollie27** [comment]
It would make sense to do this for `staticlib`s and `cydlib`s as well.

**Comment by est31** [comment]
Ideally it would be cool to go even further, to have unused lints across crates. Many crates are only used by crates inside the same project, and their pub items may or may not be used by those crates. I've built something based on save-analysis a while ago but discontinued its development. https://github.com/est31/warnalyzer

**Comment by jhpratt** [comment]
@est31 How would that work compiler-wise? My understanding is that workspaces are entirely a creation of cargo, and is no different than two unrelated crates to rustc.

**Comment by est31** [comment]
@jhpratt it's a level above the compiler, kinda like cargo-udeps. It requires save-analysis of all crates in the workspace as input. in order to support "go to places that use this", save analysis contains (thing_a -> thing_b) relationships when e.g. the thing_a function uses the thing_b function. the tool uses that info to find things that aren't used by anything else. It currently only supports one binary crate, but that's only an implementation limitation.

Anyways, the warnalyzer tool is in prototype phase, needs porting to rust-analyzer, and will always require some kind of manual user input about which crates a library crate is to be used by. Treating pub in binaries as pub(crate) can always run on the other hand, so the change should be done as well. Just wanted to make sure you are aware of warnalyzer :).

**Comment by Lokathor** [comment]
I definitely put stuff as `pub` in a binary as a deliberate way to shut up the warning about potentially dead code exiting.

**Comment by jhpratt** [comment]
@Lokathor Just curious, why? Dead code warnings are useful.

**Comment by Lokathor** [comment]
Not while things are in development, or while things depend on cargo features, or when i might want to use it later, or when

**Comment by jhpratt** [comment]
I'd be interested in trying to do this if someone can point me in the right direction.

**Comment by SorteKanin** [comment]
I'd really like this as writing `pub(crate)` everywhere in my application really hurts readability.

**Comment by estebank** [comment]
@SorteKanin you can write `crate fn foo` instead. `pub(crate)` is the 2015 edition style, but after 2018 you can write `crate` directly.

**Comment by jhpratt** [comment]
@estebank The syntax is unstable, not gated on an edition.

**Comment by estebank** [comment]
That's what happens when you spend too much time in nightly, you lose track of things. Thanks for the correction!

**Comment by yellowred** [comment]
> You can build multiple crate types at once. Then this would break any non-binary crate types that are built at the same time.

hmm but "Binaries, tests, and benchmarks are always the “bin” crate type"

**Comment by bjorn3** [comment]
> You can build multiple crate types at once. Then this would break any non-binary crate types that are built at the same time.

You can't mix a lib-like crate type with the bin crate type.

**Comment by LukasKalbertodt** [comment]
Just for those who aren't aware: this is usually not a too big problem in practice as you get "unused" warnings on `pub` items in non-`pub` modules. For example:

```rust
fn main() {}

mod bar {
    pub fn foo() {}
}
```

That warns about unused `foo`. So I'd reckon most code in the wild doesn't run into this problem. Just don't mark the top level modules as `pub`.

**Comment by mu001999** [comment]
https://github.com/rust-lang/rust/pull/149509 adds an unstable flag to treat `pub` as `pub(crate)` for binary crates

**Comment by bjorn3** [comment]
A non-trivial amount of dead code (all generic and `#[inline]` functions as well as anything small enough to be automatically marked as cross-crate inlineable) never gets codegened in the first place if it doesn't get called and as such wouldn't be caught by a link time check.

Edit: I misunderstood the proposal. I thought you meant a nicer interface to the `--print-gc-sections` option of the linker, but you mean taking the live set of symbols and then iterating over all rust functions inside the context of a rustc session and finding those which are not inside this live set. That might work if you disable MIR inlining and other optimizations that could remove functions.

**Comment by yaahc** [comment]
> I thought you meant a nicer interface to the --print-gc-sections option of the linker, 

I think this is essentially what I originally had in mind

> but you mean taking the live set of symbols and then iterating over all rust functions inside the context of a rustc session and finding those which are not inside this live set. 

This seems like a better idea

> That might work if you disable MIR inlining and other optimizations that could remove functions.

I'll keep that in mind, tyvm as always

**Comment by fmease** [comment]
On main, soon nightly, you can now pass `-Wdead_code_pub_in_binary` to basically achieve the desired behavior thanks to PR #149509 (except that it's not part of `dead_code`, of course).

**Comment by JosiahParry** [comment]
@fmease !! awesome work! How can i track when this is available via nightly? I would love to give it a try

**Comment by fmease** [comment]
I'm only the bearer of news, credit fully goes to @mu001999.

It's now on the latest nightly, that's version 2026-05-07 (rustup: nightly-2026-05-08). [Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=4234fb4f84dfe70c7b2bea5b5c537c15).

**Comment by JosiahParry** [comment]
I gave this a shot and it seems to work! But it still doesn't cover the scenario shown in https://github.com/rust-lang/rust/issues/74970#issuecomment-3429452956. nonetheless beautiful work :)

**Comment by kpreid** [comment]
I just tried out `dead_code_pub_in_binary` and, unfortunately, it interacts poorly with tests in a way that makes it difficult to use in a large workspace. In particular, if I add this to the Cargo manifest:

```rust
[lints]
rust.dead_code_pub_in_binary = "warn"
```

then I get warnings for `pub` items in my library crates any time I `cargo test` or `cargo check --tests`; I believe this is because of the test *binaries* being built from my library sources.

To fix this, I think the lint should have no effect in `rustc --test` builds. Alternatively, `cargo` could refrain from passing it to `rustc` when it’s building a test binary from a library, but that means every build system that supports library `#[test]`s needs similar logic.

**Comment by Urgau** [comment]
> I think the lint should have no effect in `rustc --test` builds

That would make `#[expect(dead_code_pub_in_binary)]` unfulfilled when built with `--test`, which is highly undesirable, as it would conflict with the normal "build" mode where it is fulfilled. 

Lints should in general never depend on the "mode" the crate is being build with.

See https://github.com/rust-lang/rust/pull/130025 where I fixed that exact issue for the `missing_docs` lint.

**Comment by mu001999** [comment]
> if I add this to the Cargo manifest:
> ```
> [lints]
> rust.dead_code_pub_in_binary = "warn"
> ```
> then I get warnings for pub items in my library crates any time I cargo test or cargo check --tests; I believe this is because of the test binaries being built from my library sources.

Yes, this can be a bit annoying in library tests at the moment. Therefore, I think the more appropriate approach for now is to enable it selectively for the target binary rather than globally. It would be better if rustc could distinguish library tests in the future.

**Comment by kpreid** [comment]
> Therefore, I think the more appropriate approach for now is to enable it selectively for the target binary rather than globally.

There isn’t just “the” binary. This lint is valuable for all kinds of targets *except for* the library target: `bin`, `test`, `bench`, and `example`. In a complex workspace with many such targets, it would be easy to accidentally skip one.

**Comment by PaulDance** [comment]
I propose moving this to a dedicated discussion in #159078, because it definitely makes the lint unusable for us as it triggers on way too much instances when enabled.
